File: Data Structures & Algorithms/trapping-rain-water/submission-1.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) {
        const n=height.length
        let trappedWater=0
        if(n<=2) return trappedWater
        let leftMax=height[0]
        let rightMax=height[n-1]
        let left=1
        let right=n-2
        while(left<=right)
        {
            if(leftMax<rightMax)
            {
                leftMax=Math.max(leftMax,height[left])
                trappedWater+=(leftMax-height[left])
                left++
            }
            else{
                rightMax=Math.max(rightMax,height[right])
                trappedWater+=(rightMax-height[right])
                right--
            }
        }
        return trappedWater
    }
}

File: Data Structures & Algorithms/trapping-rain-water/submission-1.go
func trap(height []int) int {
	n:=len(height)
	if n==0 {
		return 0
	}
	
	left:=make([]int,n)
	right:=make([]int,n)
	leftMax:=height[0]
	rightMax:=height[n-1]

	for i:=0;i<n;i++ {
		if height[i]>leftMax{
			leftMax=height[i]
		}
		left[i]=leftMax
	}
	for i:=n-1;i>=0;i--{
		if height[i]>rightMax{
			rightMax=height[i]
		}
		right[i]=rightMax
	}
	total:=0
	for i:=0;i<n;i++ {
		min:=left[i]
		if right[i]<min{
			min=right[i]
		}
		total=total+(min-height[i])
	}
	return total
}
